File: src/tools/command-queue.ts
import { z } from "zod";
import { type ToolRegistrar, jsonResult } from "./context.js";

export const registerCommandQueueTools: ToolRegistrar = (server, ctx) => {
  server.tool(
    "get_command_queue",
    "Get the due player command queue. Returns a list of players with pending " +
      "commands, the next check interval, and whether offline execution is enabled. " +
      "This is the entry point for the command delivery flow.",
    {},
    async () => {
      const data = await ctx.client.getCommandQueue();
      return jsonResult(data);
    },
  );

  server.tool(
    "get_offline_commands",
    "Get commands that can be executed immediately (offline commands). " +
      "These don't require the player to be online. Returns command id, " +
      "string, associated payment and package, and delay/slot conditions.",
    {},
    async () => {
      const data = await ctx.client.getOfflineCommands();
      return jsonResult(data);
    },
  );

  server.tool(
    "get_online_commands",
    "Get pending commands for a specific player (online commands). " +
      "These require the player to be online to execute.",
    {
      player_id: z
        .number()
        .int()
        .describe("Tebex player ID"),
    },
    async ({ player_id }) => {
      const data = await ctx.client.getOnlineCommands(player_id);
      return jsonResult(data);
    },
  );

  server.tool(
    "delete_commands",
    "Mark commands as executed and remove them from the queue. " +
      "Call this after successfully running the commands on the game server.",
    {
      ids: z
        .array(z.number().int())
        .min(1)
        .describe("Array of command IDs to remove from the queue"),
    },
    async ({ ids }) => {
      await ctx.client.deleteCommands(ids);
      return jsonResult({ ok: true, deleted_count: ids.length, ids });
    },
  );
};

File: src/tebex_mcp/tools/command_queue.py
"""Command queue tools."""

from __future__ import annotations

from typing import Annotated, Any

from fastmcp import FastMCP
from pydantic import Field

from tebex_mcp.context import ToolContext
from tebex_mcp.tools._common import IDEMPOTENT, READ_ONLY, map_tebex_errors, ok


def register(mcp: FastMCP, ctx: ToolContext) -> None:
    @mcp.tool(
        name="get_command_queue",
        description="Get due player command queue: players with pending commands, next check interval.",
        annotations=READ_ONLY,
    )
    @map_tebex_errors
    async def get_command_queue() -> Any:
        return await ctx.client.get_command_queue()

    @mcp.tool(
        name="get_offline_commands",
        description="Get commands executable without the player online: id, string, payment, package, conditions.",
        annotations=READ_ONLY,
    )
    @map_tebex_errors
    async def get_offline_commands() -> Any:
        return await ctx.client.get_offline_commands()

    @mcp.tool(
        name="get_online_commands",
        description="Get pending commands for a specific player (requires player online).",
        annotations=READ_ONLY,
    )
    @map_tebex_errors
    async def get_online_commands(
        player_id: Annotated[int, Field(description="Tebex player ID", ge=1)],
    ) -> Any:
        return await ctx.client.get_online_commands(player_id)

    @mcp.tool(
        name="delete_commands",
        description="Remove executed commands from the queue.",
        annotations=IDEMPOTENT,
    )
    @map_tebex_errors
    async def delete_commands(
        ids: Annotated[
            list[int],
            Field(description="Command IDs to remove", min_length=1),
        ],
    ) -> dict[str, Any]:
        await ctx.client.delete_commands(ids)
        return ok(deleted_count=len(ids), ids=ids)
